### PR TITLE
Add TransferableMdc

### PR DESCRIPTION
**Pull request by stIncMale**
Consider the following situation:
- in thread `t1` you received some input, decoded it in the request `r` and put some diagnostic information via `MDC`
- you want to process `r` in the some other threat `t2`, because processing takes too long and `t1` is not supposed to perform long tasks
- but you also want to have the same diagnostic information as was in the `t1` while processing `r` in the `t2`

As far as I know, current API doesn't provide such possibility (see http://stackoverflow.com/questions/6073019/how-to-use-mdc-with-thread-pools and http://shengwangi.blogspot.ru/2015/09/using-log-mdc-in-multi-thread-helloworld-example.html).

We are using `TransferableMdc` in our production servers in order to be able to use `MDC` across threads/with executors. The proposed usage is very simple and doesn't seem to has any harmful side effects:

```
TransferableMdc outerMdc = TransferableMdc.current();
executor.submit(() -> {
  try (TransferableMdc transferredMdc = outerMdc.transfer()) {
    logger.info("This call can access contents of outerMdc via org.slf4j.MDC");
  }
  logger.info("This call can not access contents of outerMdc via org.slf4j.MDC");
});
```

More examples can be found in the `TransferableMdc`'s javadoc and in `TransferableMdcTest`.

### CODE REVIEW

**Comment by girishsubramanyam** [comment]
Can this be merged ?

**Comment by ceki** [comment]
So if I understand correctly, TransferableMdc transports a copy of the map of the current thread's MDC along with the apply operation which can merge this data into a new thread.

**Comment by stIncMale** [comment]
@ceki 
Besides having the explicitly called apply method (which merges MDC's of two threads such that the transferred MDC has precedence), `TransferableMdc ` also has an implicitly called (by try-with-resources block) `close` method which restores the original MDC of the thread guaranteeing that it's MDC stays untouched.

File: slf4j-api/src/main/java/org/slf4j/TransferableMdc.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+     */
+    //@Override
+    public void close() {
+        if (context != null) {
+            MDC.clear();
+            if (backup != null) {
+                MDC.setContextMap(backup);
+                backup = null;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (stIncMale) [reply slf4j-api/src/main/java/org/slf4j/TransferableMdc.java]: Thanks @efdknittlfrank ! I fixed this and added `TransferableMdcTest.restoreEmpty` that tests the case you described.

Besides:
- added a not guaranteed check of multiple calls to `TransferableMdc.close`. Such calls signify that `TransferableMdc` is not used correctly.
- renamed `apply` to `transfer` (looks more intuitive), and changed the behaviour so that instead of merging contexts by overriding values of conflicting keys, it resets the context to the transferred state. I can't remember why I originally decided to merge contexts instead. Does anyone have ideas on what behaviour is preferable: merge vs reset? It's also possible to support both and let the user decide, I am just not sure how useful it would be.

[the remark this answers is not in the document]
Comment (ceki) [reply slf4j-api/src/main/java/org/slf4j/TransferableMdc.java]: In my opinion, restoring the MDC to its old state does not makes sense. 

In the case of servlet filters which can be cascaded, it makes sense for the latest filter not to clobber its predecessors when starting. Similarly, when exiting it should only clear keys that it has set itself, leaving other keys alone. However, if it has set a given key, I don't see a use case where it makes sense to restore that key to its old value.

[the remark this answers is not in the document]
Comment (stIncMale) [reply slf4j-api/src/main/java/org/slf4j/TransferableMdc.java]: @ceki I am not sure I understand the situation you described. Specifically, I don't what is specific about asynchronous handling of requests in filters/servlets comparing to asynchronous handling of other tasks:

- The majority of projects use synchronous processing in `Filter.doFilter` and `Servlet.service`, and in such a situations `TransrefableMdc` is not required (there is no other thread a user may want to transfer the current MDC to). For the situations when `Filter.doFilter`/`Servlet.service` handle requests asynchronously by starting async processing via `ServletRequest.startAsync`, the `FilterChain.doFilter` is still called sequentially. So also no place for `TransrefableMdc`.

- As far a I understand, the only situation when `TransrefableMdc` can be used with filters or servlets, is when after calling `ServletRequest.startAsync` a user submits an async task either via `AsyncContext.start` or by calling `ExecutorService.submit` on a user-controlled executor service (or something similar), but in this case it's not different from submitting any task to any executor.

If you still see something special about servlet filters, could you please describe the situation in more details (maybe a code snippet?).

The idea of `TransrefableMdc.close` is simple: execute a task and leave thread's state (MDC) exactly as it was before executing the task. Why? - So that other tasks don't observe any changes made not by them. Imagine a general-purpose executor (e.g. `ForkJoinPool.commonPool`, but it can be anything). I may want to transfer MDC when submitting a task there. Someone else (someone else's code) may want to just execute the task knowing that MDC is supposed to be empty because he never messed with it in executor's threads. However since `TransrefableMdc` does not rollback MDC changes (what you are proposing?), the second user may now observe unexpected changes made to MDC previously by my task.

[the remark this answers is not in the document]
Comment (efdknittlfrank) [reply slf4j-api/src/main/java/org/slf4j/TransferableMdc.java]: The way I see the behavior of `TransferableMdc` is that it creates a "Stack" of MDCs. Each `apply` (now `transferAnd…`) operation, pushes to the stack, and each `close` pops the stack, restoring the previous value.
In my eyes, that's sensible behavior (I'm currently relying on it)

[the remark this answers is not in the document]
Comment (stIncMale) [reply slf4j-api/src/main/java/org/slf4j/TransferableMdc.java]: While we can interpret `TransferableMdc ` as a stack where you can push (`transfer`) not more than one element, and you can pop (`close`) not more than one element. I want to propose a different view of it. But before that, let's consider the situation you probably had in mind when imagining a stack: multiple levels of "nesting".

**Nesting**
For multiple levels of asynchronous nesting we can write the following terrible code:
```
TransferableMdc outermostMdc = TransferableMdc.current();
executor1.submit(() -> {
  try (TransferableMdc transferredMdc = outermostMdc.transfer()) {
    TransferableMdc outerMdc = TransferableMdc.current();
    executor2.submit(() -> {
        try (TransferableMdc weNeedToGoDeeperMdc = outerMdc.transfer()) {
            //is this deep enough?
        }
    }
  }
});
```
I agree, that multiple levels of asynchronous nesting may happen in practice, but the code above is not how it should be handled. A better code would look like the following (the example is artificial, but I believe shows the point):
```
void process(Request req) {
  //...
  submitMdcAware(executor1, () -> processFirstPart(req));
}

void processFirstPart(Request req) {
  //...
  submitMdcAware(executor2, () -> processSecondPart(req));
}

void processSecondPart(req) {
  //...
}

void submitMdcAware(ExecutorService executor, Runnable action) {
  TransferableMdc outerMdc = TransferableMdc.current();
  executor.submit(() -> {
    try (TransferableMdc transferredMdc = outerMdc.transfer()) {
      action.run();
    }
  });
}
```
So the code may be using `TransferableMdc` without even knowing it (see `process` and `processFirstPart`), and multiple levels of asynchrony are handled trivially without supporting any stack functionality in the `TransferableMdc`.

**A (correct? :) view on `TransferableMdc` which is different from seeing it as a stack**
There is no problem in transferring MDC between threads. In fact, both Java's inner classes and lambda expressions capture effectively final variables from their context, thus making the transfer of MDC between threads automatic (as soon as you have a copy of it's map and you access it from a lambda expression). What we are missing is the ability to apply the transferred context (that's why I called the method `apply` originally) and roll it back once it's not needed anymore. And this is what `TransferableMcd` does for a user in a most compact way I can see possible in Java (if someone sees what the library can do to make to code shorter without using proxies or code generation, please share).

P.S. Speaking of proxies, it's definitely possible to for example create `MdcAwareExecutorService implements ExecutorService` (it's method `submit` would look similar to `submitMdcAware` above), so that a user would not need to do anything and MDC would be transferred, applied and rolled back transparently. But SL4J is not a right place for such a class, while `TransferableMcd` provides a building block naturally belonging to SL4J that can be used by others e.g. for building such MDC-aware wrappers.

[the remark this answers is not in the document]
Comment (ceki) [reply slf4j-api/src/main/java/org/slf4j/TransferableMdc.java]: The reason I mentioned servlet filters was to illustrate a case of component composition where each component sets part of the MDC key space. In such a configuration, either the "first" component calls MDC.clear() either at the very beginning or at the very end of processing. In the more general case where the "first" filter cannot be identified, an alternative strategy is for each component to clear the keys it set. In both approaches the idea is to avoid leaving stale data in the thread which might pollute the view of subsequent callers. 

**In light of this concern (i.e. not to leave stale data behind), the idea of restoring previously existing (and most probably stale data) seems quite counterproductive.**

[the remark this answers is not in the document]
Comment (stIncMale) [reply slf4j-api/src/main/java/org/slf4j/TransferableMdc.java]: @ceki I believe I get what you are saying. Does the following example captures your idea?
```
TransferableMdc outerMdc = TransferableMdc.current();
  executor.submit(() -> {
    /*inner MDC already has the "numberOfSubmittedTasks" key
      initially set to 0 by a thread factory*/
    try (TransferableMdc transferredMdc = outerMdc.transferAndMerge()) {
      MDC.put("numberOfSubmittedTasks",//increment the counter
          Integer.parseInt(MDC.get("numberOfSubmittedTasks")) + 1);
    }
    /*here we still want want to see the incremented value of "numberOfSubmittedTasks"
      so that the next submitted task could also increment it's value,
      instead of always incrementing 0 as it would happen
      if TransferableMdc.close() were rolling back all MDC changes*/
});
}
```

**Merge/unmerge**
I agree that if `TransferableMdc` merges the transferred MDC with the inner MDC via `transferAndMerge`, the behaviour of `close` should be as you mentioned: remove new keys introduced by merging, leave all other keys (and their potentially updated values) that were in the inner MDC before merging (essentially, it's an unmerge operation). Do you agree with this?

**Override/rollback**
The following question still needs to be answered: what should the behaviour of `close` be when `transferAndOverride` is used? It appears to me that in this case rollback (reinstate the original context) is the correct behaviour. Thoughts?

**A request of example from @ceki**
You also mentioned "_a case of component composition where each component sets part of the MDC key space_". If you still see this as something possible with multiple threads involved, could you please give a specific example (not necessary filter/servlet, just anything artificial)? I am asking because I don't think something like this is even possible in theory (or I don't understand your idea; a specific example will help in either case).

[the remark this answers is not in the document]
Comment (ceki) [reply slf4j-api/src/main/java/org/slf4j/TransferableMdc.java]: My concern is quite unsophisticated. I updated my earlier comment to shorten the discussion.

[the remark this answers is not in the document]
Comment (stIncMale) [reply slf4j-api/src/main/java/org/slf4j/TransferableMdc.java]: @ceki Thanks for the example. Let's see what we have there.

**The third task**
> // this message should not bear data for "client" but probably will

This is true because MDC is thread-scoped, and the only thing that can be done in order to exclude the possibility of observing the `"client"` key is explicitly removing it from MDC (e.g. by calling `MDC.clear`either at the end of the first task or the beginning of the third task). This situation has nothing to do with `TransferableMdc` and the quoted statement is true regardless of the presence of the second task (the one using `TransferableMdc`).

**The second task**
> // this message should not bear MDC data for "client"

If we want this behaviour, we should use `transferAndOverride` instead of `transferAndMerge`. That's it.

To sum up, the problem in the third task has nothing to do with `TransferableMdc`, and the problem in the second task is solved with `TransferableMdc`. But interestingly, none of the problems is about the behaviour of `TransferableMdc.close`.

[the remark this answers is not in the document]
Comment (stIncMale) [reply slf4j-api/src/main/java/org/slf4j/TransferableMdc.java]: @ceki Do we still have unresolved concerns about either the behaviour of `close` or the proposed merge/unmerge, override/rollback functionality? If not, I will start implementing that. If yes, then lets continue figuring those out.

[the remark this answers is not in the document]
Comment (efdknittlfrank) [reply slf4j-api/src/main/java/org/slf4j/TransferableMdc.java]: To leave my 2¢: I would very much depend on the "rollback" (stacklike) behavior of close. After leaving a try-with-resources block, I expect to contain the MDC the same keys and values as it did before the block, regardless of any updates that might have happened inside of the block. (Very much like the `putCloseable` on MDC, but not being limited to a single item and supporting multiple key-value pairs at a time)

[the remark this answers is not in the document]
Comment (ceki) [reply slf4j-api/src/main/java/org/slf4j/TransferableMdc.java]: The problem TransferableMdc.close() does not help with the stale data problem. Moreover, it is not hard to imagine situations where it makes it worse. As I mentioned previously, I don't see the point in restoring potentially stale data. Apparently it is useful for nesting MDCs but why would anyone want nested MDCs? Anyone care to elaborate?

[the remark this answers is not in the document]
Comment (efdknittlfrank) [reply slf4j-api/src/main/java/org/slf4j/TransferableMdc.java]: Our use case is as follows: Often we have several MDC values which should only be set for a single log entry. All log messages are logged to Elasticsearch, in which each MDC value is a separate property of the document. This allows easy filtering on or visualization of certain log properties.
In such a case, we do not want to lose the original value of the MDC key (which potentially is non-empty). After writing the log, the MDC should be restored to the original values, not cleared fully (because some values should really apply to _all_ log messages). Nesting MDCs simplifies this pattern, without having to manually remember the old values (Thinking about it, storing a copy and restoring it manually could also work in this case?).
Maybe we are abusing the MDC here, but it is a super simple and effective way to enable our logs to be sliced and diced (and visualized). There's StructuredArguments, but they have some shortcomings too.

[the remark this answers is not in the document]
Comment (stIncMale) [reply slf4j-api/src/main/java/org/slf4j/TransferableMdc.java]: > why would anyone want nested MDCs? Anyone care to elaborate?
For us the situation was as follows:
- receive a new HTTP request
- discover client's address from the request and add it to MDC so that we can see it in log messages related to handling the request
- pass the request into a different thread pool for processing (IO threads do not precess requests)
`TransferableMdc` is used to not loose client's address related to the request as a result of handling the request in a thread different from the one received the request and set the address into MDC.

Can the aforementioned be done without `TransferableMdc`? Sure it can, but we can also do anything we do with `MDC` without having `MDC` in SLF4J. However having `MDC` makes things easier. The same is true for `TransferableMdc`.

> The current implementation of TransferableMdc.close() does not help with the stale data problem.
It's not suppose to help with this problem. This problem exists in `MDC` regardless of the presence of `TransferableMdc`, and can be solved e.g. by calling the `clear` method.

> Moreover, it is not hard to imagine situations where it makes it worse.
It seems like it's hard to imagine: I can't see such a situation and you did not present a plausible example. If you don't agree with my analysis of the example you provided, please point out what specifically you disagree about.

I am claiming that if `close` leaves the state of `MDC` exactly the same as it was before `TransferableMdc` was applied, then it's not possible for external code (the code executed outside of the block where `TransferableMdc` is applier) to observe the fact that `TransferableMdc` is used.

[the remark this answers is not in the document]
Comment (ceki) [reply slf4j-api/src/main/java/org/slf4j/TransferableMdc.java]: The need to transferring MDC from one thread to another is **not** being contested. The argument is about the appropriateness of TransferableMdc.close where it **restores** previous data.  

There is no issue explicitly supporting nested MDC by an API which makes the nested structure explicit. The current proposal sneaks in the nested behavior. I am sorry to say that just bad API design.

[the remark this answers is not in the document]
Comment (stIncMale) [reply slf4j-api/src/main/java/org/slf4j/TransferableMdc.java]: > that just bad API design and not something I can get behind

I can't disagree or agree with a qualitative statement unless it is accompanied by arguments or I can see the arguments on my own. In this case, I can't see the arguments on my own, so I keep asking you to give an example where the introduction of `TransferableMdc` and specifically the call to `TransferableMdc.close` (which restores the original MDC state) makes the possible behaviours worse than they were before the introduction of `TransferableMdc`.

> it is not hard to imagine situations where it makes it worse: a developer clears the MDC only to have it restored by TransferableMdc.close()

May I see a specific example that you have in mind when describing this situation? Just communicate this idea as code so that I don't have to think out (and potentially make mistakes) of what you think of. And once we have a specific argument expressed in a strict form (code), we will be able to discuss it in a productive manner.

**Comment by ceki** [comment]
Here is a copy and paste from a previous comment https://github.com/qos-ch/slf4j/pull/150/commits/25ea1e0f7167fec52ee4f924bb04a0d42b0a3d0b#r307683110

I only added MDC.clear() in the second task submission.

```// in some part of the application which does not use 
  // nor is aware of TransferableMdc
  executor.submit(() -> {
     // assume the task is executed within one of the threads
     // in a thread pool. Let us refer to this thread as P0
		MDC.put("client", "YN-1000-B");
  }); 
	
  // at a later time, in some totally unrelated thread,  a different part 
  // of the application invokes  the following code
  MDC.put("key", "value");
  TransferableMdc outerMdc = TransferableMdc.current();
  executor.submit(() -> {
  	// assume the second task is also delegated to P0
  	try (TransferableMdc transferredMdc = outerMdc.transferAndMerge()) {
  		logger.debug("a  message");
                
  		// this message should not bear MDC data for "client"
  	} **finally {
            MDC.clear();
        }**
  });

  // a third task created at a later time also delegated to P0
  executor.submit(() -> {
  	// this message should not bear data for "client" but probably will
  	logger.debug("some other message");
  }); 
```

**Comment by stIncMale** [comment]
I am assuming that "_this message should not bear data for "client" but probably will_" in the third task is the specific problem you are referring to.

Executions of all the tasks in the example are concurrent, i.e. there is no total order between them. This means that even if we remove `TransferableMdc` from the example completely (while leaving the call to `MDC.clear()` in the second task), the third task will still me able to observe the `MDC` state written by the first task:
```java
executor.submit(() -> {
  MDC.put("client", "YN-1000-B");
}); 

MDC.put("key", "value");
executor.submit(() -> {
  MDC.clear();
});

executor.submit(() -> {
  //this message still may bear the data for "client" despite TransferableMdc is not involved
  logger.debug("some other message");
}); 
```
Since the third task may observe the data written by the first task regardless of the presence of `TransferableMdc`, it's true that `TransferableMdc` did not made the situation worse (in fact, it had no effect on the possible outcomes).

Now we may say that `executor` is a single-thread executor, and then execution of all tasks is definitely ordered (i.e. not concurrent). It's true, but the order is not constrained in the code (and therefore not known in advance) if submissions happen concurrently from different threads. This again give us the same results as with multi-threaded `executor`: the third task may observe the writes from the first task even without `TransferableMdc` being involved despite the second task clears the context.

We can go even further and say that not only the `executor` is single-threaded, but also submission of all tasks happened from a single thread (or was somehow else totally ordered with a predefined order) in the order first -> second -> third. **In the situation when execution of all tasks are totally ordered (first -> second -> third) it's true that if the second task clears `MDC`, the third task must not observe the data written by the first one.** And again this behaviour stays unchanged once `TransferableMdc` is introduced the way you specified:
```java
try (TransferableMdc transferredMdc = outerMdc.transferAndMerge()) {
  ...
}  finally {
  //MDC.clear is called after TransferableMdc.close as specified in https://docs.oracle.com/javase/specs/jls/se12/html/jls-14.html#jls-14.20.3.2
  MDC.clear();
}
```
because the second task clears `MDC` after `TransferableMdc` is closed.

I believe I considered all relevant possibilities, and in none of them `TransferableMdc` had any affect on the possibility of observing the `MDC` state left by the first task from the second task. Therefore, unless you can show where I am incorrect, or what I have missed, we should conclude that the proposed behaviour of `close` does not make things worse (alternatively, you may come up with another example).

**Comment by ceki** [comment]
By later time, I meant really later, as in perhaps 10 hours later, corresponding to a **recycled** thread by the Executor implementation. 

As for the "finally" block being executed last per the JLS, as a developer I should not need to worry about such a detail just for transferring MDC data from a parent to a child thread. Moreover, I could have just as easily written

```java
 executor.submit(() -> {
  	// assume the second task is also delegated to P0
  	try (TransferableMdc transferredMdc = outerMdc.transferAndMerge()) {
  		logger.debug("a  message");              
  		// this message should not bear MDC data for "client"

               // as good citizens we clear the MDC    
              MDC.clear();  
        }
  });
```

In short, TransferableMdc.close() introduces too much baggage where it is not necessarily needed.

Note that nested MDC might be very useful in certain circumstances. For such circumstances, it should be supported explicitly with a distinct data structure.

**Comment by stIncMale** [comment]
> By later time, I meant ... a recycled thread

In this case the situation is analogous to a single-threaded executor with all submits happened from the same thread (with the difference that what you are describing is just one of many allowed executions).

@ceki Just to recap:

1. You think that `TransferableMdc ` that removes signs of its presence via the `close` method (such that the MDC stays in exactly the same state as it was before using `TransferableMdc`) makes things worse.
2. As an example of a situation where things get worse you are using code with a piece that pollutes MDC. You want to get rid of the pollution. So the code is already incorrect in a way, and one of the solutions is to fix the code that pollutes MDC. But maybe you don't have control over that polluting code, that's a possibility, so we are proceeding.
3. Instead of getting rid of MDC pollution the way it should be done, i.e. by clearing MDC at the beginning of the task that wants MDC to be clean, the example is doing this at the end of a different _unrelated_ task.
4. When it was shown (despite of the aforementioned problem) that introducing `TransferableMdc` to such code allows one to have the desired behaviour, you counter-argue by saying that one may not know the basics of Java syntax (specifically try-with-resources behaviour, which is very simple), and one may write the try-finally block incorrectly (I can't take arguments like this one seriously).

I don't see a possibility for us to come to an agreement. Since you are the one making decisions about the project, I am removing myself from the discussion about the behaviour of the method `TransferableMdc.close` so that you could proceed the way you see correct (e.g. by modifying my code and including it in the project, or by rejecting it and closing the pull request, etc.).

**Comment by ceki** [comment]
@stIncMale Notwithstanding the disagreement about `TransferableMdc.close`, I thank you for your contribution. and the present conversation. Both are much appreciated.

**Comment by stIncMale** [comment]
@ceki Sure, thank you for developing and maintaining extremely useful logging libraries 👍

**Comment by Hollerweger** [comment]
I see issues with MDC not being populated in a spring service using parallelStreams. Could I safely use TransferableMdc instead to solve the issue?

**Comment by efdknittlfrank** [comment]
@Hollerweger yes, should be possible. Since operation of a parallelStream run on different threads than the calling thread, the MDC is lost (MDC is thread local). The MDC needs to be copied to each operation (and later cleared/restored to not leak thread local information). TLDR; yes, `TransferableMdc` should help you transfer+restore the other thread's MDC.

**Comment by Hollerweger** [comment]
Thanks @efdknittlfrank 
@ceki Would be great if this PR get's merged (or similar solution) as all existing solutions found on the Web to solve the issue are quite awful. E.g.: https://stackoverflow.com/a/57772891/1100559
Devs tend to forget to copy the MDC and a more generic solution like this one is much appreciated.

**Comment by Hollerweger** [comment]
@efdknittlfrank Seems I misunderstood the concept. There is still a manual copy operation required for each new Thread.
Would it be possible to have a solution that automatically copies MDC to all new Threads (maybe with an exception for a thread name pattern)?

**Comment by efdknittlfrank** [comment]
@Hollerweger `TransferableMdc` is used as follows:

```java
final TransferableMdc mdc = TransferableMdc.current();
collection.parallelStream().map(item -> {
  try (Mdc unused = mdc.apply()) {
    return yourMappingLogicHere(item);
  }
}).collect(toList());
```

each function (`item -> … `) will be a closure over the value of `mdc`. Functions will be picked up by the thread pool and each function then applies the outer MDC (variable `mdc`) to the currently executing thread. The try-with-resources guarantees that the thread's original MDC is restored after the function completed.

There is no "magic" to automatically apply MDC to child tasks, unless you count custom thread factories or wrappers:

```java
private static <T, R> Function<T, R> transferMdc(final Function<T, R> mapper) {
  // executed in outer/parent thread
  final TransferableMdc mdc = TransferableMdc.current();

  // closure, executed in thread pool
  return item -> {
    try (Mdc unused = mdc.apply()) {
      return yourMappingLogicHere(item);
    }
  };
}

final TransferableMdc mdc = TransferableMdc.current();
collection.parallelStream().map(applyMdc(item -> yourMappingLogicHere(item)).collect(toList());
```

Note that the MDC is captured from the thread where the map operation is defined, not when the stream is collected.

**Comment by stIncMale** [comment]
@Hollerweger
> Would it be possible to have a solution that automatically copies MDC?

A generic solution like that is impossible in principle. Specific solutions that will do the job automatically are, however, possible. Following is an example of a `Stream` decorator.
```java
/**
 * An {@link Stream} decorator that {@linkplain TransferableMdc transfers}
 * {@link MDC} between threads when needed.
 * This decorator is useful when using {@linkplain Stream#isParallel() Parallel} streams.
 */
public class MdcAwareStream<T> implements Stream<T> {
  private final Stream<T> delegate;

  public MdcAwareStream(Stream<T> delegate) {
    this.delegate = delegate;
  }

  @Override
  public <R> Stream<R> map(Function<? super T, ? extends R> mapper) {
    final TransferableMdc outerMdc = TransferableMdc.current();
    return delegate.map(t -> {
      try (var transferredMdc = outerMdc.transfer()) {
        return mapper.apply(t);
      }
    });
  }

  //do the same as above for all Stream methods with parameters that may contain user code (params of type Predicate, Supplier, Function, etc.)

  @Override
  public long count() {
    return delegate.count();
  }

  //implement other trivial methods similarly to the above
}
```
Now you can decorate any parallel `Stream` so that it automatically copies MDC:

```java
MDC.put("key", "value");
List<Integer> vals = Stream.generate(() -> ThreadLocalRandom.current().nextInt(1, Integer.MAX_VALUE))
    .limit(10_000)
    .collect(Collectors.toList());
List<Integer> halved = new MdcAwareStream<>(vals.parallelStream())
    .map(val -> {
      assert "value".equals(MDC.get("key"));//"key" is accessible despite the mapping function may be executed by multiple threads
      return val / 2;
    })
    .collect(Collectors.toList());
```
